compiler/modules/front/resolveTypes.bal

```diff
@@ -28,6 +28,41 @@ function resolveTypes(ModuleSymbols mod) returns ResolveTypeError? {
     foreach var defn in mod.defns {
         check resolveDefn(mod, defn);
     }
+    check finishDeferredEmptinessChecks(mod);
+}
+
+function finishDeferredEmptinessChecks(ModuleSymbols mod) returns ResolveTypeError? {
+    if !mod.tc.env.isReady() {
+        panic err:impossible("type environment is not ready");
+    }
+    foreach var { semType, modDefn, td } in mod.deferredEmptinessChecks {
+        // XXX When we can give multiple errors we should check all the deferred emptiness checks
+        check finishEmptinessCheck(mod, semType, modDefn, td);
+    }
+    foreach var [checked, loc] in mod.recursiveEmptinessChecked {
+        if !checked {
+            return err:semantic("non recursive empty type", loc);
+        }
+    }
+}
+
+function finishEmptinessCheck(ModuleSymbols mod, t:SemType semType, s:ModuleLevelDefn modDefn, s:TypeDesc td) returns ResolveTypeError? {
+    if t:isEmpty(mod.tc, semType) {
+        d:Location loc = s:locationInDefn(modDefn, { startPos: td.startPos, endPos: td.endPos });
+        if td is s:BinaryTypeDesc && td.op is "&" {
+            return err:semantic("intersection must not be empty", loc);
+        }
+        var [isRecursive, index] = t:isSemTypeRecursive(semType);
+        if isRecursive {
+            if index is int {
+                mod.recursiveEmptinessChecked[index.toString()] = [true, loc];
```

Comment:
You should use a table rather than a map with int:toString.

Reply:
Fixed